clash-prelude/src/Clash/Sized/Vector.hs

```diff
@@ -1058,8 +1058,26 @@ foldr f z (x `Cons` xs) = f x (foldr f z xs)
 -- delay, of O(@'length' xs@). Use 'fold' if your binary operator @f@ is
 -- associative, as @"'fold' f xs"@ produces a structure with a depth of
 -- O(log_2(@'length' xs@)).
-foldl :: (b -> a -> b) -> b -> Vec n a -> b
-foldl f z xs = last (scanl f z xs)
+foldl :: forall b a n . (b -> a -> b) -> b -> Vec n a -> b
+foldl f z0 xs0
+  -- We use 'go' to make 'foldl' strict during Haskell simulation, while using
+  -- 'scanl' to make Clash's life easier during normalization. We make 'foldl'
+  -- strict due to similar reasoning to:
+  --
+  --   https://github.com/hasura/graphql-engine/pull/2933#discussion_r328821960
+  --
+  -- Also see:
+  --
+  --   https://well-typed.com/blog/90/
+  --
+  | clashSimulation = go z0 xs0
+  | otherwise = last (scanl f z0 xs0)
+ where
+  go :: forall m. b -> Vec m a -> b
+  go z Nil = z
+  go z (Cons x xs) =
+    let z1 = f z x
+    in z1 `seq` go z1 xs
```

Comment:
```suggestion
    in z1 `seqX` go z1 xs
```

?

Reply:
`seqX` would allow using a function that is lazy in its first argument to keep working as they did before `foldl` became strict, but it is also pretty confusing to people comfortable with bottom values that not all bottoms are treated equal in a function explicitly marked as being strict. So I'm inclined to treat all bottoms as producing bottom by definition.

Reply:
Another option, a `seq` where exceptions in its first argument are simply ignored, would at least treat all bottoms as equal but still violate the expectance that a strict `foldl` would in fact return bottom whenever a bottom is encountered while accumulating. So, once again, I'm inclined to use `seq`.

Reply:
Maybe we should offer a variant with the behaviour that bottoms don't propagate up, so people can still actually write a function lazy in its first argument producing results.

Reply:
I'm inclined to use `seqX`, for the reasons you mention:

> `seqX` would allow using a function that is lazy in its first argument to keep working for `XException` as it did before `foldl` became strict,

Because hunting `XException`s is a pretty miserable experience a lot of the time.

Reply:
We would already differ in the respect that `foldl` is strict _at all_, so I suppose if we clearly document `foldl` as being strict with the exception of ignoring `XException`s in its accumulator, that would also work. But we do need the extra documentation because what we're doing is somewhat strange.

Reply:
That's fair, but I think that's only because we specialize its types. As mentioned in the documentation of `Data.Foldable`, it never makes sense to use `foldl` on lists. We're somewhat special because we (ab)use bottoms to represent X.

Also I thought newer `base`s already make `foldl` strict for `List`s, but it doesn't seem to be so:

```haskell
foldl k z0 xs =
  foldr (\(v::a) (fn::b->b) -> oneShot (\(z::b) -> fn (k z v))) (id :: b -> b) xs z0
```

(Also if you think that implementation is unreadable, look at the one in `Foldable`:

```haskell
foldl f z t = appEndo (getDual (foldMap (Dual . Endo . flip f) t)) z
```

)

Reply:
Where does it say it doesn't make sense on lists? They even mention the case we're discussing here as an application:
> This is rarely what you want, but can work well for structures with efficient right-to-left sequencing and an operator that is lazy in its left argument.

Reply:
Is there a good reason for propagating other bottoms than `XException`? Couldn't we just ignore all bottoms while forcing the accumulator, and have it outwardly appear to work as the lazy version? As soon as the function actually forces its first argument to WHNF, the bottom would still pop out.

Reply:
I mean that's pretty heavily implied.. In any case, [lexi lambda's](https://github.com/hasura/graphql-engine/pull/2933#discussion_r328821960) comment makes the case pretty convincingly. Unless there is some glaring mistake in her reasoning, I don't think we should rehash the discussion here.

Reply:
1. Did you answer to a message I already deleted because it wasn't that important / I changed my mind?
2. I think the case is convincingly made. but we are discussing doing something different. We're discussing ignoring certain bottom values. She doesn't say anything about that, so I don't see what we're rehashing.

Reply:
Yes! I did respond to the deleted comment. I'll delete mine too (and this one after :-)).

Reply:
In what case would we even observe the difference? The function would be really weird... where it ignores its arguments in some cases, and return exceptions in others...

```
f 0 y = errorX "QQ"
f _ 2 = 1
f x y = x + y

foldl f 1 [(-1),2]
```
would return `1` if we used `seqX`, and raise an exception if we used `seq`.

I would be nice to have some benchmark numbers to determine whether the performance difference is negligible enough to use `seqX` over `seq`.

Reply:
With

```haskell
{-# LANGUAGE BangPatterns, DataKinds, NoImplicitPrelude, GADTs, ScopedTypeVariables, NumericUnderscores #-}
module Main where

import Clash.Prelude
import Clash.Sized.Vector
import Control.DeepSeq

rnfY :: NFData a => Vec n a -> ()
rnfY = foldl (\() -> rnf) ()

main :: IO ()
main = do
  let !_ = rnfY (repeat @100_000_000 (0 :: Int))
  return ()
```

and

```
cabal build clash-prelude && ghc -fforce-recomp Test.hs
```

and

```
time ./Test && time ./Test && time ./Test && time ./Test && time ./Test && time ./Test
```

I get:

| seq   | seqX  | diff   |
|-------|-------|--------|
| 2.932 | 5.082 | 57.69% |
| 2.932 | 5.152 | 56.91% |
| 2.902 | 5.152 | 56.33% |
| 2.942 | 5.192 | 56.66% |
| 2.932 | 5.182 | 56.58% |
| 2.932 | 5.252 | 55.83% |